### PR TITLE
ADOP 210 change 

### PR DESCRIPTION
**Pull request by pankajhmcts**
### Change description
Implementation of first two screen.

### JIRA link
https://tools.hmcts.net/jira/browse/ADOP-210

**Does this PR introduce a breaking change?** (check one with "x")

- [ ] Yes
- [X ] No

### CODE REVIEW

File: src/main/steps/children/adoption-agency/content.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import { FormContent, FormFieldsFn } from '../../../app/form/Form';
+import { isEmailValid, isFieldFilledIn, isPhoneNoValid } from '../../../app/form/validation';
+
+export const en = (): Record<string, unknown> => ({
+  section: 'Your adoption agency or local authority details',
+  title: 'Adoption agency or local authority details',
+  adopAgencyName: 'Name of adoption agency or local authority',
+  adopAgencyPhone: 'Phone number',
+  adopAgencyContactName: 'Name of your contact',
+  adopAgencyEmail: 'Email address of your contact',
```
Comment (abidfs) [inline src/main/steps/children/adoption-agency/content.ts:11]:
```suggestion
  adopAgencyContactEmail: 'Email address of your contact',
```

File: src/main/steps/children/adoption-agency/content.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  },
+});
+
+export const cy = (): Record<string, unknown> => ({
+  section: 'Your adoption agency or local authority details (in Welsh)',
+  title: 'Adoption agency or local authority details (in Welsh)',
+  adopAgencyName: 'Name of adoption agency or local authority (in Welsh)',
+  adopAgencyPhone: 'Phone number (in Welsh)',
+  adopAgencyContactName: 'Name of your contact (in Welsh)',
+  adopAgencyEmail: 'Email address of your contact (in Welsh)',
```
Comment (abidfs) [inline src/main/steps/children/adoption-agency/content.ts:36]:
```suggestion
  adopAgencyContactEmail: 'Email address of your contact (in Welsh)',
```

File: src/main/steps/children/adoption-agency/get.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import { GetController } from '../../../app/controller/GetController';
+
+@autobind
+export default class AdoptionAgencyGetController extends GetController {
+  public async get(req: AppRequest, res: Response): Promise<void> {
+    const adopAgencyOrLAs = req.session.userCase.adopAgencyOrLAs || [];
+
+    let redirect = false;
+    if (req.query.add) {
+      console.log('req.query.add 14: ' + req.query.add);
```
Comment (abidfs) [inline src/main/steps/children/adoption-agency/get.ts:14]: Remove unwanted console statements

File: src/main/steps/children/adoption-agency/get.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      req.session.userCase.selectedAdoptionAgencyId = `${req.query.add}`;
+      req.session.userCase.hasAnotherAdopAgencyOrLA = undefined;
+      delete req.query.add;
+      req.url = req.url.substring(0, req.url.indexOf('?'));
+      redirect = true;
+    }
+    //else if (req.query.change) {
```
Comment (abidfs) [inline src/main/steps/children/adoption-agency/get.ts:21]: remove commented code

File: src/main/steps/children/adoption-agency/get.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
+import autobind from 'autobind-decorator';
```
Comment (abidfs) [inline src/main/steps/children/adoption-agency/get.ts:1]: Unit test not covered for this file

File: src/main/steps/children/adoption-agency/post.test.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
+import { assert } from 'console';
+
+describe('post content', () => {
+  it('should return the correct content for language = en', () => {
+    assert(true);
+  });
+});
+
+// const mockGetParsedBody = jest.fn();
```
Comment (abidfs) [inline src/main/steps/children/adoption-agency/post.test.ts:9]: Unit tests missing

File: src/main/steps/children/childrenSequence.test.ts [hunk 7]
```diff
@@ -3,8 +3,8 @@ import { YesOrNo } from '../../app/case/definition';
 import { childrenSequence } from './childrenSequence';
 
 describe('childrenSequence', () => {
-  test('should contain 11 entries in children screen sequence', () => {
-    expect(childrenSequence).toHaveLength(11);
+  test('should contain 13 entries in children screen sequence', () => {
+    expect(childrenSequence).toHaveLength(13);
```
Comment (abidfs) [inline src/main/steps/children/childrenSequence.test.ts:7]: missing expect for newly added objects in sequence

File: src/main/steps/children/other-adoption-agency/content.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
+import { YesOrNo } from '../../../app/case/definition';
+import { TranslationFn } from '../../../app/controller/GetController';
+import { FormContent } from '../../../app/form/Form';
+import { isFieldFilledIn } from '../../../app/form/validation';
+
+const en = () => ({
+  section: 'Your adoption agency or local authority details',
+  title: 'Was there another adoption agency or local authority involved in placing the child?',
+  yes: 'Yes',
+  no: 'No',
```
Comment (abidfs) [inline src/main/steps/children/other-adoption-agency/content.ts:10]: yes and no are already declared in common content, you can remove it

File: src/main/steps/children/other-adoption-agency/content.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+      required: 'Please answer the question',
+    },
+  },
+});
+
+const cy = () => ({
+  section: 'Your adoption agency or local authority details (in Welsh)',
+  title: 'Was there another adoption agency or local authority involved in placing the child? (in Welsh)',
+  yes: 'Yes (in Welsh)',
+  no: 'No (in Welsh)',
```
Comment (abidfs) [inline src/main/steps/children/other-adoption-agency/content.ts:22]: yes and no are already declared in common content, you can remove it

File: src/main/steps/children/other-adoption-agency/content.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
+import { YesOrNo } from '../../../app/case/definition';
+import { TranslationFn } from '../../../app/controller/GetController';
+import { FormContent } from '../../../app/form/Form';
+import { isFieldFilledIn } from '../../../app/form/validation';
+
+const en = () => ({
+  section: 'Your adoption agency or local authority details',
+  title: 'Was there another adoption agency or local authority involved in placing the child?',
```
Comment (abidfs) [inline src/main/steps/children/other-adoption-agency/content.ts:8]: Please make it label
```suggestion
  label: 'Was there another adoption agency or local authority involved in placing the child?',
```

File: src/main/steps/children/other-adoption-agency/content.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  errors: {
+    hasAnotherAdopAgencyOrLA: {
+      required: 'Please answer the question',
+    },
+  },
+});
+
+const cy = () => ({
+  section: 'Your adoption agency or local authority details (in Welsh)',
+  title: 'Was there another adoption agency or local authority involved in placing the child? (in Welsh)',
```
Comment (abidfs) [inline src/main/steps/children/other-adoption-agency/content.ts:20]: same as english

**Review by abidfs — CHANGES_REQUESTED** [review]
looks good but few minor comments

File: src/main/steps/children/adoption-agency/get.test.ts [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+      expect(req.session.userCase.adopAgencyOrLAs).toEqual([{ adopAgencyOrLaId: '1609459200000' }]);
+    });
+  });
+
+  describe('when there is a adoptionAgency with selectedAdoptionAgencyId in userCase', () => {
+    test('should not create a blank adoptionAgency', async () => {
+      req = mockRequest({
+        session: {
+          userCase: {
+            selectedAdoptionAgencyId: 'MOCK_PLACEMENT_ORDER_ID',
```
Comment (abidfs) [inline src/main/steps/children/adoption-agency/get.test.ts:58]: Can you please remove placement order references from this file

File: src/main/steps/children/adoption-agency/get.ts [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+import { GetController } from '../../../app/controller/GetController';
+
+@autobind
+export default class AdoptionAgencyGetController extends GetController {
+  public async get(req: AppRequest, res: Response): Promise<void> {
+    const adopAgencyOrLAs = req.session.userCase.adopAgencyOrLAs || [];
+
+    let redirect = false;
+    if (req.query.add) {
+      console.log('req.query.add 14: ' + req.query.add);
```
Comment (abidfs) [inline src/main/steps/children/adoption-agency/get.ts:14]: I see you resolved the conversation but did not remove the console log

File: src/main/steps/children/adoption-agency/get.ts [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+      req.url = req.url.substring(0, req.url.indexOf('?'));
+      redirect = true;
+    } else if (
+      !req.session.userCase.selectedAdoptionAgencyId ||
+      req.session.userCase.selectedAdoptionAgencyId === 'undefined'
+    ) {
+      //generate random id for adoption agency if there are no adoption agency
+      req.session.userCase.selectedAdoptionAgencyId = adopAgencyOrLAs[0]?.adopAgencyOrLaId || `${Date.now()}`;
+    }
+    console.log('get 38 ' + req.session.userCase.selectedAdoptionAgencyId);
```
Comment (abidfs) [inline src/main/steps/children/adoption-agency/get.ts:32]: same here, please remove console log

File: src/main/steps/children/childrenSequence.test.ts [hunk 15]
```diff
@@ -53,5 +53,15 @@ describe('childrenSequence', () => {
     expect(childrenSequence[10].url).toBe('/children/full-name-after-adoption');
     expect(childrenSequence[10].showInSection).toBe('aboutChildren');
     expect(childrenSequence[10].getNextStep({})).toBe('/task-list');
+
+    expect(childrenSequence[11].url).toBe('/children/adoption-agency');
+    expect(childrenSequence[11].showInSection).toBe('aboutChildren');
+    expect(
+      childrenSequence[11].getNextStep({ adopAgencyOrLAs: [{ adopAgencyOrLaId: '1' }, { adopAgencyOrLaId: '2' }] })
+    ).toBe('/children/other-adoption-agency');
```
Comment (abidfs) [inline src/main/steps/children/childrenSequence.test.ts:61]: need to cover other scenarios as well

File: src/main/steps/children/childrenSequence.test.ts [hunk 16]
```diff
@@ -53,5 +53,15 @@ describe('childrenSequence', () => {
...
+
+    expect(childrenSequence[12].url).toBe('/children/other-adoption-agency');
+    expect(childrenSequence[12].showInSection).toBe('aboutChildren');
+    expect(childrenSequence[12].getNextStep({ hasAnotherAdopAgencyOrLA: YesOrNo.NO })).toBe('/task-list');
```
Comment (abidfs) [inline src/main/steps/children/childrenSequence.test.ts:65]: need to cover other scenarios as well
see the jest coverage report for more details

File: src/main/steps/children/childrenSequence.ts [hunk 17]
```diff
@@ -61,4 +61,25 @@ export const childrenSequence: Step[] = [
...
+  {
+    url: Urls.ADOPTION_AGENCY,
+    showInSection: Sections.AboutChildren,
+    getNextStep: userCase =>
+      userCase.adopAgencyOrLAs!.length < 2
+        ? Urls.OTHER_ADOPTION_AGENCY
+        : userCase.adopAgencyOrLAs?.length === 2 &&
+          userCase.selectedAdoptionAgencyId === userCase.adopAgencyOrLAs[1].adopAgencyOrLaId
+        ? Urls.TASK_LIST_URL
+        : Urls.OTHER_ADOPTION_AGENCY,
```
Comment (abidfs) [inline src/main/steps/children/childrenSequence.ts:73]: unit tests not covered for all scenarios

File: src/main/steps/children/childrenSequence.ts [hunk 18]
```diff
@@ -61,4 +61,25 @@ export const childrenSequence: Step[] = [
...
+  },
+  {
+    url: Urls.OTHER_ADOPTION_AGENCY,
+    showInSection: Sections.AboutChildren,
+    getNextStep: userCase =>
+      userCase.hasAnotherAdopAgencyOrLA === YesOrNo.NO
+        ? Urls.TASK_LIST_URL
+        : userCase.adopAgencyOrLAs !== undefined && userCase.adopAgencyOrLAs?.length < 2
+        ? `${Urls.ADOPTION_AGENCY}?add=${Date.now()}`
+        : `${Urls.ADOPTION_AGENCY}?change=${userCase.adopAgencyOrLAs![1].adopAgencyOrLaId}`,
```
Comment (abidfs) [inline src/main/steps/children/childrenSequence.ts:83]: unit tests not covered for all scenarios

File: src/main/steps/task-list/content.ts [hunk 19]
```diff
@@ -25,6 +26,14 @@ const getSectionStatusLabel = (status, statuses, id) => {
   }
 };
 
+const getAdoptionAgencyUrl = userCase => {
+  if (userCase.adopAgencyOrLAs?.length) {
+    const adopAgency = userCase.adopAgencyOrLAs[0];
+    return `${urls.ADOPTION_AGENCY}?change=${adopAgency.adopAgencyOrLaId}`;
+  }
+  return `${urls.ADOPTION_AGENCY}?add=${Date.now()}`;
+};
```
Comment (abidfs) [inline src/main/steps/task-list/content.ts:35]: as discussed, please move this method to task-list > utils.ts file
also cover the unit tests for the same

File: src/test/e2e/pages/taskList.page.js [hunk 20]
```diff
@@ -23,6 +23,8 @@ module.exports = {
     childOtherParentDetailsStatus: 'strong[id$="other-parent-status"]',
     childrenBirthFatherLink: 'a[id$="children-birth-father"]',
     childrenBirthFatherDetailsStatus: 'strong[id$="birth-father-status"]',
+    adoptionAgencyLink: 'a[id$="adoption-agency"]',
+    adoptionAgencyDetailsStatus: 'strong[id$="birth-father-status"]',
```
Comment (abidfs) [inline src/test/e2e/pages/taskList.page.js:27]: I think the id is wrong it shouldn't be "birth-father-status"

**Review by abidfs — CHANGES_REQUESTED** [review]
few minor changes needed

**Review by abidfs — DISMISSED** [review]
Approving it but I think `getAdoptionAgencyDetailStatus` logic can be improved
up to you

**Review by nraghu32 — APPROVED** [review]

**Pull request merged**